pytorch_lightning/core/lightning.py

```diff
@@ -1757,3 +1700,54 @@ def get_tqdm_dict(self) -> Dict[str, Union[int, str]]:
         rank_zero_warn("`get_tqdm_dict` was renamed to `get_progress_bar_dict` in v0.7.3"
                        " and this method will be removed in v1.0.0", DeprecationWarning)
         return self.get_progress_bar_dict()
+
+    def _auto_register_hparams(self):
+        """
+        Removes the need to pass in hparams. Instead, we register every argument in init
+        to the module with some caveats:
+        1. we don't overwrite the property if it already exists
+        2. we also store a module_arguments property for model loading and saving
+        """
+        # two frames back is the init of the child module
+        frame = inspect.currentframe()
+        # todo: this shall be flexible to find all init in the path, recursion?
+        frame_args = frame.f_back.f_back.f_locals
+
+        frame_args = _collect_init_args(frame)
+        init_args = {k: v for k, v in frame_args.items()
+                     if k not in ('args', 'kwargs', 'self', '__class__', 'frame', 'frame_args')}
+        child = _get_latest_child(frame)
+
+        # we'll save hparams automatically (renamed to module_arguments)
+        for arg, val in init_args.items():
+            # don't overwrite something already set
+            if hasattr(child, arg):
+                continue
+            setattr(child, arg, val)
+
+        # set module_arguments in child
+        setattr(child, 'module_arguments', init_args)
+
+
+def _collect_init_args(frame, args: dict = {}) -> dict:
+    """Recursive search for all children."""
```

Comment:
I don't think we want to recurse through all children and collect all of their init arguments. We just want to save the arguments for the init method that actually gets called explicitly in the code right?

Reply:
and what about default kwargs? we cannot just ignore them...

Reply:
what I have seen (correct me if I am wrong) it is not possible to distinguish from `kwarg` given directly and `kwarg` taken from declaration with default value...

Reply:
i don't think we need to distinguish between kwargs and default kwargs. we just need to save whatever arguments were given to the very first call to the init function, default value or not.

Reply:
```python
class MyModule(pl.LightningModule):
    def __init__(self, my_arg, my_kwarg=1):
        ....
class Child(MyModule):
    def __init__(self, child_arg, child_kwarg='hi'):
         ...
class GrandChild(Child):
    def __init__(self. grand_child_arg, grand_child_kwarg=1.2):
        ...

MyModule(1)  # we want to save {'my_arg': 1} (my_kwarg will be filled in automatically by Python)
MyModule(1, my_kwarg=2)  # we want to save {'my_arg': 1, 'my_kwarg': 2}

Child(3) # we want to save {'child_arg': 3} (child_kwarg will be filled in automatically by Python)
Child(3, child_kwarg='hey')  # we want to save {'child_arg': 3, 'child_kwarg': 'hey'}

GrandChild(5)  # we want to save {'grand_child_arg': 5} (grand_child_kwarg will be filled in automatically by Python)
GrandChild(5, grand_child_kwarg=2.5)  # we want to save {'grand_child_arg': 5, 'grand_child_kwarg': 2.5}
```

This is what I mean when I said we should get to the leaf init function that gets called explicitly in the code.

Reply:
I am not sure about ignoring arguments of parents here, it is like you would rewrite all methods in child class (which you would not do because it is against the inheritance principles)
cc: @williamFalcon

Reply:
ahh i see what you mean. yeah this would break when a child class doesn't implement its own `__init__()`.

This again makes me go back to my (and @tullie's) original proposal of adding an abstract property, and have people explicitly specify what needs to be passed into the constructor when restoring. Yeah it's another thing to implement, but I do think it provides the cleanest implementation.

Reply:
well actually, would it break? I don't think so b/c if the child class doesn't have an `__init__()` defined, Python would just call the init of its parent, which would be the leaf frame, and we just need to save that. I think my new approach discussed here, and my original approach of adding a property both work.

Reply:
we have a test case with an empty init too

Reply:
@yukw777 do you want to add the tests for the cases you're concerned about? i want to make sure we explicitly support/don't break those.